tests/functional/ArchiveTest.php

```diff
@@ -0,0 +1,427 @@
+                    'db-host' => $this->fixtureDatabaseSettings['db-hostname'],
+                    'db-user' => $this->fixtureDatabaseSettings['db-username'],
+                    'db-password' => $this->fixtureDatabaseSettings['db-password'],
+                ]
+            )
+        );
+        $this->assertRestoredSiteStatus();
+
+        // Restore database with a set of database connection options with an invalid host.
+        $this->drush(
+            'archive:restore',
+            [],
+            array_merge(
+                array_diff_key($this->archiveRestoreOptions, ['db-url' => null]),
+                [
+                    'db' => null,
+                    'db-source-path' => Path::join($this->extractPath, 'database', 'database.sql'),
+                    'db-driver' => $this->fixtureDatabaseSettings['db-driver'],
+                    'db-name' => $this->fixtureDatabaseSettings['db-name'],
+                    'db-host' => 'invalid_host',
+                    'db-user' => $this->fixtureDatabaseSettings['db-username'],
+                    'db-password' => $this->fixtureDatabaseSettings['db-password'],
+                ]
+            ),
+            null,
+            null,
+            self::EXIT_ERROR
+        );
+        $this->assertStringContainsString(
+            sprintf('Failed to create database %s.', $this->fixtureDatabaseSettings['db-name']),
+            $this->getErrorOutput()
+        );
+        $this->assertRestoredSiteStatus();
+
+        // Restore archive from a non-existing file.
+        $nonExistingArchivePath = Path::join($this->getSandbox(), 'arch.tar.gz');
+        $this->drush(
+            'archive:restore',
+            [$nonExistingArchivePath],
+            $this->archiveRestoreOptions,
+            null,
+            null,
+            self::EXIT_ERROR
+        );
+        $this->assertStringContainsString(
+            'arch.tar.gz is not found',
+            $this->getErrorOutput()
+        );
+        $this->assertRestoredSiteStatus();
+
+        // Restore database without database connection settings.
+        $this->drush(
+            'archive:restore',
+            [],
+            array_merge(
+                array_diff_key($this->archiveRestoreOptions, ['db-url' => null]),
+                [
+                    'db' => null,
+                    'db-source-path' => Path::join($this->extractPath, 'database', 'database.sql'),
+                ]
+            ),
+            null,
+            null,
+            self::EXIT_ERROR
+        );
+        $this->assertStringContainsString(
+            'Database connection settings are required if --destination-path',
+            $this->getErrorOutput()
+        );
+        $this->assertRestoredSiteStatus();
+    }
+
+    /**
+     * Asserts the status of restored site.
+     */
+    private function assertRestoredSiteStatus(): void
+    {
+        $this->setupSettingsPhp();
```

Comment:
The tests shouldn't have to write the db credentials to settings.php; `archive:restore` should do it.